doc/rieman.1
.TH rieman 1 "2023-02-27" "GNU/Linux" "User's Manual"

.SH "NAME"
rieman \- The X11 desktop pager

.SH "SYNOPSIS"
.B rieman
[
.B \-h
]
[
.B \-v[v]
]
[
.B \-c
config
]
[
.B \-w
]
[
.B \-l logfile
]

.SH "DESCRIPTION"
.PP
.B rieman
is X11 desktop pager: a program that display each existing virtual
desktop with its windows and allows to switch between them.

.SS "FEATURES"
.IP \(bu 4
NetWM compliant
.IP \(bu 4
Large desktop/Viewports support
.IP \(bu 4
Ability to set desktops geometry (_NET_DESKTOP_GEOMETRY) to arrange
desktops in a configurable 2D grid
.IP \(bu 4
Displays name of desktops and windows in a configurable info area
.IP \(bu 4
Displays desktop number
.IP \(bu 4
Window selection/focusing when switching
.IP \(bu 4
Configurable look for different window states (active/inactive/requires attention)
.IP \(bu 4
Mini per-desktop tray for hidden windows
.IP \(bu 4
Supports display of application window icons
.IP \(bu 4
Pseudo-transparency support
.IP \(bu 4
Theme (aka skin) support. Any element may be customized.
.IP \(bu 4
Configuration reload
.IP \(bu 4
Simple tiling for desktop windows (1.1)
.IP \(bu 4
Control commands via configuration socket (1.2)
.IP \(bu 4
Per-output desktop subset display (1.3)


.SH "OPTIONS"
.TP
.B \-help
Displays short message with command-line options synopsis.
.TP
.B \-v
displays version

.TP
.B \-vv
displays version and build-time configuration

.TP
.B \-w
.B rieman
starts up in withdrawn mode so it can go into slit (aka dock). Exact
effect is defined by a running window manager.

.TP
.B \-c <config>
use alternate
.I config
file. This overrides configuration lookup scheme.

.TP
.B \-l <logfile>
use
.I logfile
instead of sdout/stderr

.SH "SIGNALS"
.PP

To reload configuration, send
.B SIGUSR1
to the rieman process. It will re-read all configuration files and apply
changes.

.SH "ENVIRONMENT VARIABLES"
.PP

The used X11 screen is taken from the
.B DISPLAY
environment variable.  To run rieman on the second screen run:
.TP
    $ DISPLAY=:0.1 rieman

.SH "FILES"

.PP
The main configuration files is named rieman.conf.
A skin is a directory containing image files and a
rieman_skin.conf configuration file. All paths inside
skin configuration file are relative to its directory.

Files are looked up until found in the following locations:

.TP
.I ./conf/rieman.conf
.TP
.I ./skins/<name>/rieman_skin.conf

current directory configuration (i.e. executing from the build directory)

.TP
.I
$XDG_CONFIG_HOME/rieman/rieman.conf
.TP
.I $XDG_DATA_HOME/rieman/skins/<name>/rieman_skin.conf

location as defined by freedesktop.org

.TP
.I ~/.config/rieman.conf
.TP
.I ~/.local/share/rieman/skins/<name>/rieman_skin.conf

home directory

.TP
.I /usr/local/share/rieman/rieman.conf
.TP
.I /usr/local/share/rieman/skins/<name>/rieman_skin.conf

system-wide, /usr/local/share/rieman is the default datadir

.SH "CONFIGURATION"

Configuration file format:

.IP \(bu 4
everything from '#' to end of line is a comment and ignored

.IP \(bu 4
each line contains key and value separated by whitespace

.IP \(bu 4
value is not escaped and quoting is not required. leading and trailing
spaces in the value are ignored. See conf/rieman.conf for example.

Configuration keys are:

.TP
.I geometry.width, geometry.height <0 | n>

Defines size of a single on-screen desktop cell in pixels. Zero height leads
to autocalculated setting based on screen aspect ratio.

.TP
.I layout.wrap <n>
   layout.corner <c>
   layout.orientation <horizontal | vertical>

The wrap parameter defines length of a row or column (depending on orientation).
The corner is one of "topleft", "topright", "bottomleft" or "bottomright" and
defines how desktops are arranged. See _NET_DESKTOP_LAYOUT for details.

.TP
.I subset.enabled <true | false>

.I subset.output <RANDR_output>

.I subset.start_desktop <n>

.I subset.ndesktops <n>

The "subsets" mode enables special mode, in which:

1) only a subset of virtual desktops is shown (start...start + ndesktops).

2) desktop geometry is determined by specified RandR output resolution.

This is useful in multi-monitor configuration, where each monitor has its
own set of virtual desktops, and desktop size is a summ of monitor sizes.
The mode allows to run multiple instance of rieman, each attached to
specific monitor and showing only desktops visible on the specific monitor.

The feature was tested with awesome WM. Note that root window property
_NET_CURRENT_DESKTOP allows to be set only once, so you cannot have active
desktop highlighted in all rieman instances (in awesome's terms, this is
currently selected tag, to which we have on access from outside).

.TP
.I appearance.skin <name>

Sets the skin name to use.

.TP
.I appearance.desktop_text <true | false>
   appearance.desktop_text.content <name|number>

If set, desktop number or name is displayed in the center of a desktop cell

.TP
.I appearance.desktop_pad <true | false>
   appearance.desktop_pad.position <above | below>

If set, little pad is drawn above or below desktop cell, which is used
to display desktop name (and for other purposes as well, like windows
names and icons for hidden windows)

.TP
.I appearance.window_icon <true | false>

If set, window icon is displayed additionally in a window rectangle

.TP
.I appearance.viewports  <true | false>

If set, viewports are displayed as a grid inside a desktop cell

.TP
.I appearance.minitray  <true | false>

If set, icons of a hidden windows are displayed in area, where desktop
name is shown; click on such an icon restores the window

.TP
.I window.withdrawn  <true | false>
    starts the pager in withdrawn state to put into dock/slit

.TP
.I window.dock  <true | false>
    sets window type to _NET_WM_WINDOW_TYPE_DOCK.
    Useful for WMs that expect it.

.TP
.I window.skip_taskbar  <true | false>

    If set, pager window is not shown in the taskbar

.TP
.I window.skip_pager  <true | false>

    If set, pager window is not shown in pagers

.TP
.I window.sticky  <true | false>

    If set, pager window appears on all desktops

.TP
.I window.position <topright | topleft | bottomright | bottomleft>
   window.dx <n>
   window.dy <n>

    Defines startup position on the screen. Note that in withdrawn mode
    window may be reparented and the setting will define position inside
    dock

    Optional dx and dy attributes control offset from x/y border at specified
    corner. By default, offsets are zero.

.TP
.I window.layer  <above | below | normal>

    Defines window layer

.TP
.I window.struts  <true | false>

.I window.struts.left <pixels>

.I window.struts.left_start_y <pixels>

.I window.struts.left_end_y <pixels>

.I window.struts.right right_start_y <pixels>

.I window.struts.right_end_y <pixels>

.I window.struts.top top_start_x <pixels>

.I window.struts.top_end_x <pixels>

.I window.struts.bottom <pixels>

.I window.struts.bottom_start_x <pixels>

.I window.struts.bottom_end <pixels>

   If enabled, defines values for_NET_WM_STRUT/_NET_WM_STRUT_PARTIAL which
   reserve place on screen borders for the window.

.TP
.I actions.change_desktop  <true | false>
   actions.change_desktop.mouse_button <left | right | middle>

    If set, mouse events are processed, allowing to switch desktops and focus
    windows. The mouse_button arguments selects the desired button.

.TP
.I actions.tile_windows <true | false>
   actions.tile_windows.mouse_button <left | right | middle>

    If set, mouse events are processed, allowing to tile windows on clicked
    desktop, using different tiling methods in a loop. Currently supported
    method is "fair" ported from the awesome WM, in horizontal and vertical modes.
    The mouse_button arguments selects the desired button.

.TP
.I control socket </path/to/socket>

   If set, rieman will listen for commands on UNIX DGRAM socket at specified
   path.  To send command, run the rieman executable with -m option and provide
   path to socket and comand, for example:

.I  $ rieman -m /path/to/ctl.sock exit

.SS "Recognized commands"

.IP \(bu 4
reload - reload configuration, see description above
.IP \(bu 4
exit - terminate the process
.IP \(bu 4
switch_desktop_left/right/up/down - switch desktop in 2D grid coordinates
.IP \(bu 4
switch_desktop_prev/next - switch to prev/next desktop by order
.IP \(bu 4
tile_current_desktop - tile windows on current desktop, one layout per command


.SH "SKIN CONFIGURATION"

The skin configuration defines following elements:
.IP \(bu 4
fonts - font specification for pager elements
.IP \(bu 4
backgrounds - textures specifications for pager elements
.IP \(bu 4
borders - borders specification for pager elements
.IP \(bu 4
icons - icons settings

.TP
The font attributes are:
.TP
.I fonts.<item>.face <freetype_def>
   fonts.<item>.size <n>
   fonts.<item>.color <[0x]base16-value>

.I face
is passed to fontconfig, check corresponding docs. Example "Droid Sans:10:Bold".
.I size
is the actual size in pixels that will be used to show it on screen.
.I color
is a hex number.

.TP
The backgrounds attributes are:
.TP
.I backgrounds.<item>.type <image | color>
   backgrounds.<item>.alpha <0..1>
   backgrounds.<item>.src <path | :root:>

CHANGES
version 1.2.0: 2021-01-03

    * changed configuration file format due to removing libxml dependency
      now files use '.conf' extension and use simple key-value syntax.
      use scripts/migrate_to_conf.sh to convert existing configuration files
      into new format

    + Tiling support: for WMs lacking it
      Since rieman knows about all windows and can send commands to them,
      it was simple to add 'tile windows on click at specific desktop'
      feature. See man for details.

    + Control socket support
      you can send commands to running instance via UNIX socket.
      Everything that could be done with signals and more.
      More commands are expected in future versions.
      Why? Not all WMs can deal with 2D-desktops grid and provide commands
      to switch desktop in horizontal/vertical direction.

    + NET_WM_STRUT family of atoms are supported. Rieman may now be configured
      to reserve space on screen in compatible WMs.
      Now it is also possible to set window type to 'dock'.
      Window position handling is improved.

    + X11 screen to use is now configurable

    + Root window geometry is now tracked

    + Psedo-transparency works properly with multiple monitors

    * Lot of various small bugfixes

version 1.1.0: was never released
    * version skipped due to breaking changes introduced

version 1.0.0: 2017-12-26
    + bugfixes and cleanup

version 0.9.1: 2017-12-24
    * first released version


src/rieman.c

/*
 * Copyright (C) 2017-2023 Vladimir Homutov
 */

/*
 * This file is part of Rieman.
 *
 * Rieman is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Rieman is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 */

#include "rieman.h"
#include "rie_event.h"
#include "rie_config.h"
#include "rie_xcb.h"
#include "rie_skin.h"
#include "rie_external.h"

#include <stdio.h>
#include <limits.h>

#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <signal.h>


static rie_conf_map_t rie_conf_layers[] = {
    { "below", RIE_WIN_STATE_BELOW },
    { "normal", 0 },
    { "above", RIE_WIN_STATE_ABOVE },
    { NULL, 0 }
};

static rie_conf_map_t rie_conf_orientations[] = {
    { "horizontal", XCB_EWMH_WM_ORIENTATION_HORZ },
    { "vertical", XCB_EWMH_WM_ORIENTATION_VERT },
    { NULL, 0 }
};

static rie_conf_map_t rie_conf_positions[] = {
    { "topleft", RIE_POS_TOPLEFT },
    { "topright", RIE_POS_TOPRIGHT },
    { "bottomleft", RIE_POS_BOTTOMLEFT },
    { "bottomright", RIE_POS_BOTTOMRIGHT },
    { NULL, 0 }
};

static rie_conf_map_t rie_conf_corners[] = {
    { "topleft", XCB_EWMH_WM_TOPLEFT },
    { "topright", XCB_EWMH_WM_TOPRIGHT },
    { "bottomleft", XCB_EWMH_WM_BOTTOMLEFT },
    { "bottomright", XCB_EWMH_WM_BOTTOMRIGHT },
    { NULL, 0 }
};

static rie_conf_map_t rie_conf_pad_positions[] = {
    { "below", RIE_PAD_POS_BELOW },
    { "above", RIE_PAD_POS_ABOVE },
    { NULL, 0 }
};

static rie_conf_map_t rie_conf_buttons[] = {
    { "left", XCB_BUTTON_MASK_1 },
    { "middle", XCB_BUTTON_MASK_2 },
    { "right", XCB_BUTTON_MASK_3 },
    { NULL, 0 }
};

static rie_conf_map_t rie_conf_labels[] = {
    { "number", RIE_DESKTOP_NUMBER },
    { "name",  RIE_DESKTOP_NAME },
    { NULL, 0 }
};


static rie_conf_item_t rie_conf[] = {

    { "geometry.width", RIE_CTYPE_UINT32, "100",
      offsetof(rie_settings_t, desktop.w), NULL, { NULL } },

    { "geometry.height", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, desktop.h), NULL, { NULL } },

    { "layout.wrap", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, desktop.wrap), NULL, { NULL } },

    { "layout.corner", RIE_CTYPE_STR, "topleft",
      offsetof(rie_settings_t, desktop.corner),
      rie_conf_set_variants, { &rie_conf_corners } },

    { "layout.orientation", RIE_CTYPE_STR, "horizontal",
      offsetof(rie_settings_t, desktop.orientation),
      rie_conf_set_variants, { &rie_conf_orientations } },


    { "appearance.skin", RIE_CTYPE_STR,  "default",
      offsetof(rie_settings_t, skin), NULL, { NULL } },

    { "appearance.desktop_pad", RIE_CTYPE_BOOL, "true",
      offsetof(rie_settings_t, show_pad), NULL, { NULL } },

   { "appearance.desktop_pad.position", RIE_CTYPE_STR, "below",
      offsetof(rie_settings_t, pad_position),
      rie_conf_set_variants, { &rie_conf_pad_positions } },

   { "appearance.desktop_pad.margin", RIE_CTYPE_UINT32, "2",
      offsetof(rie_settings_t, pad_margin), NULL, { NULL } },

    { "appearance.desktop_text", RIE_CTYPE_BOOL, "true",
      offsetof(rie_settings_t, show_text), NULL, { NULL } },

    { "appearance.desktop_text.content", RIE_CTYPE_STR, "number",
      offsetof(rie_settings_t, desktop.content),
      rie_conf_set_variants, { &rie_conf_labels } },

    { "appearance.window_icon", RIE_CTYPE_BOOL, "true",
      offsetof(rie_settings_t, show_window_icons), NULL, { NULL } },

    { "appearance.viewports", RIE_CTYPE_BOOL, "true",
      offsetof(rie_settings_t, show_viewports), NULL, { NULL } },

    { "appearance.minitray", RIE_CTYPE_BOOL, "true",
      offsetof(rie_settings_t, show_minitray), NULL, { NULL } },

    { "window.withdrawn", RIE_CTYPE_BOOL, "false",
      offsetof(rie_settings_t, withdrawn), NULL, { NULL } },

    { "window.skip_taskbar", RIE_CTYPE_BOOL, "true",
      offsetof(rie_settings_t, skip_taskbar), NULL, { NULL } },

    { "window.skip_pager", RIE_CTYPE_BOOL, "true",
      offsetof(rie_settings_t, skip_pager), NULL, { NULL } },

    { "window.sticky", RIE_CTYPE_BOOL, "true",
      offsetof(rie_settings_t, sticky), NULL, { NULL } },

    { "window.layer", RIE_CTYPE_STR, "normal",
      offsetof(rie_settings_t, layer),
      rie_conf_set_variants, { &rie_conf_layers } },

    { "window.position", RIE_CTYPE_STR, "topleft",
      offsetof(rie_settings_t, position),
      rie_conf_set_variants, { &rie_conf_positions } },


    { "actions.change_desktop.mouse_button", RIE_CTYPE_STR, "left",
      offsetof(rie_settings_t, change_desktop_button),
      rie_conf_set_variants, { &rie_conf_buttons } },

    /* configuration schema 1.1 */

    { "actions.change_desktop", RIE_CTYPE_BOOL, "true",
      offsetof(rie_settings_t, enable_change_desktop_button), NULL, { NULL } },

    { "actions.tile_windows", RIE_CTYPE_BOOL, "true",
      offsetof(rie_settings_t, enable_tile_button), NULL, { NULL } },

    { "actions.tile_windows.mouse_button", RIE_CTYPE_STR, "right",
      offsetof(rie_settings_t, tile_button),
      rie_conf_set_variants, { &rie_conf_buttons } },

    /* configuration schema 1.2 */

    { "window.dock", RIE_CTYPE_BOOL, "false",
      offsetof(rie_settings_t, docked), NULL, { NULL } },

    { "window.strut", RIE_CTYPE_BOOL, "false",
      offsetof(rie_settings_t, struts.enabled), NULL, { NULL } },

    { "window.strut.left", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, struts.left), NULL, { NULL } },

    { "window.strut.left_start_y", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, struts.left_start_y), NULL, { NULL } },

    { "window.strut.left_end_y", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, struts.left_end_y), NULL, { NULL } },

    { "window.strut.right", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, struts.right), NULL, { NULL } },

    { "window.strut.right_start_y", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, struts.right_start_y), NULL, { NULL } },

    { "window.strut.right_end_y", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, struts.right_end_y), NULL, { NULL } },

    { "window.strut.top", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, struts.top), NULL, { NULL } },

    { "window.strut.top_start_x", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, struts.top_start_x), NULL, { NULL } },

    { "window.strut.top_end_x", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, struts.top_end_x), NULL, { NULL } },

    { "window.strut.bottom", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, struts.bottom), NULL, { NULL } },

    { "window.strut.bottom_start_x", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, struts.bottom_start_x), NULL, { NULL } },

    { "window.strut.bottom_end_x", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, struts.bottom_end_x), NULL, { NULL } },

    { "control.socket", RIE_CTYPE_STR, "",
      offsetof(rie_settings_t, control_socket_path), NULL, { NULL } },

    { "window.position.dx", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, pos_x_offset), NULL, { NULL } },

    { "window.position.dy", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, pos_y_offset), NULL, { NULL } },

    /* configuration schema 1.3 */

    { "subset.enabled", RIE_CTYPE_BOOL, "false",
      offsetof(rie_settings_t, subset.enabled), NULL, { NULL } },

    { "subset.output", RIE_CTYPE_STR, "",
      offsetof(rie_settings_t, subset.output), NULL, { NULL } },

    { "subset.start_desktop", RIE_CTYPE_UINT32, "0",
      offsetof(rie_settings_t, subset.start_desktop), NULL, { NULL } },

    { "subset.ndesktops", RIE_CTYPE_UINT32, "1",
      offsetof(rie_settings_t, subset.ndesktops), NULL, { NULL } },

    { NULL, 0, NULL, 0, NULL, { NULL } }
};


#if defined(RIE_TESTS)
rie_t    *pagerp;
#endif

uint8_t   rie_withdraw;
uint8_t   rie_quit;
uint8_t   rie_reload;


int
rie_pager_init(rie_t *pager, rie_t *oldpager)
{
    pager->selected_desktop = -1;
    pager->selected_vp.x = -1;
    pager->selected_vp.y = -1;
    pager->m_x = -1;
    pager->m_y = -1;

    if (oldpager) {
        pager->xcb = oldpager->xcb;
        pager->log = oldpager->log;

    } else {
        pager->xcb = rie_xcb_new(pager->cfg);
        if (pager->xcb == NULL) {
            return RIE_ERROR;
        }
    }

    if (rie_xcb_set_desktop_layout(pager->xcb, pager->cfg) != RIE_OK) {
        return RIE_ERROR;
    }

    pager->gfx = rie_gfx_new(pager->xcb);
    if (pager->gfx == NULL) {
        return RIE_ERROR;
    }

    pager->skin = rie_skin_new(pager->cfg->skin, pager->gfx);
    if (pager->skin == NULL) {
        return RIE_ERROR;
    }

    if (rie_event_init(pager) != RIE_OK) {
        return RIE_ERROR;
    }

    if (rie_xcb_set_window_hints(pager->xcb, pager->cfg, pager->current_desktop)
        != RIE_OK)
    {
        return RIE_ERROR;
    }

    if (strlen(pager->cfg->control_socket_path)) {
        pager->ctl = rie_control_new(pager->cfg, pager);
        if (pager->ctl == NULL) {
            return RIE_ERROR;
        }
    }

    return RIE_OK;
}


static void
rie_switch_desktop(rie_t *pager, rie_command_t direction)
{
    rie_desktop_t  *desk;
    int32_t         left, right, up, down, curr;

    desk = pager->desktops.data;

    curr = pager->current_desktop;
    desk = &desk[curr];

    left = right = up = down = 0;

    if (direction == RIE_CMD_SWITCH_DESKTOP_PREV) {
        if (curr == 0) {
            return;
        }
        rie_xcb_set_desktop(pager->xcb, curr - 1);
        return;
    }

    if (direction == RIE_CMD_SWITCH_DESKTOP_NEXT) {
        if (curr == pager->desktops.nitems - 1) {
            return;
        }
        rie_xcb_set_desktop(pager->xcb, curr + 1);
        return;
    }

    if (pager->cfg->desktop.orientation == XCB_EWMH_WM_ORIENTATION_HORZ) {

        switch (pager->cfg->desktop.corner) {
        case XCB_EWMH_WM_TOPLEFT:
        case XCB_EWMH_WM_BOTTOMLEFT:
            left = curr - 1;
            right = curr + 1;
            break;

        case XCB_EWMH_WM_TOPRIGHT:
        case XCB_EWMH_WM_BOTTOMRIGHT:
            left = curr + 1;
            right = curr - 1;
            break;
        default:
            break;
        }

        switch (pager->cfg->desktop.corner) {
        case XCB_EWMH_WM_TOPLEFT:
        case XCB_EWMH_WM_TOPRIGHT:
            up = curr - pager->ncols;
            down = curr + pager->ncols;
            break;

        case XCB_EWMH_WM_BOTTOMRIGHT:
        case XCB_EWMH_WM_BOTTOMLEFT:
            up = curr + pager->ncols;
            down = curr - pager->ncols;
            break;
        default:
            break;
        }

    } else { /* XCB_EWMH_WM_ORIENTATION_VERT */

        switch (pager->cfg->desktop.corner) {
        case XCB_EWMH_WM_TOPLEFT:
        case XCB_EWMH_WM_TOPRIGHT:
            up = curr - 1;
            down = curr + 1;
            break;

        case XCB_EWMH_WM_BOTTOMRIGHT:
        case XCB_EWMH_WM_BOTTOMLEFT:
            up = curr + 1;
            down = curr - 1;
            break;
        default:
            break;
        }

        switch (pager->cfg->desktop.corner) {
        case XCB_EWMH_WM_TOPLEFT:
        case XCB_EWMH_WM_BOTTOMLEFT:
            left = curr - pager->nrows;
            right = curr + pager->nrows;
            break;

        case XCB_EWMH_WM_TOPRIGHT:
        case XCB_EWMH_WM_BOTTOMRIGHT:
            left = curr + pager->nrows;
            right = curr - pager->nrows;
            break;
        default:
            break;
        }
    }

    switch (direction) {
    case RIE_CMD_SWITCH_DESKTOP_LEFT:
        if (desk->lcol == 0) {
            return;
        }
        rie_xcb_set_desktop(pager->xcb, left);

        break;
    case RIE_CMD_SWITCH_DESKTOP_RIGHT:
        if (desk->lcol == pager->ncols - 1) {
            return;
        }
        rie_xcb_set_desktop(pager->xcb, right);

    break;

    case RIE_CMD_SWITCH_DESKTOP_UP:
        if (desk->lrow == 0) {
            return;
        }
        rie_xcb_set_desktop(pager->xcb, up);

        break;
    case RIE_CMD_SWITCH_DESKTOP_DOWN:
        if (desk->lrow == pager->nrows - 1) {
            return;
        }
        rie_xcb_set_desktop(pager->xcb, down);

        break;

    default:
        /* make compiler happy with incomplete enum switch */
        break;
    }
}


void
rie_pager_run_cmd(rie_t *pager, rie_command_t cmd)
{
    switch (cmd) {

    case RIE_CMD_RELOAD:
        rie_reload = 1; /* to be handled by event loop */
        break;

    case RIE_CMD_EXIT:
        rie_quit = 1; /* to be handled by event loop */
        break;

    case RIE_CMD_SWITCH_DESKTOP_LEFT:
    case RIE_CMD_SWITCH_DESKTOP_RIGHT:
    case RIE_CMD_SWITCH_DESKTOP_UP:
    case RIE_CMD_SWITCH_DESKTOP_DOWN:
    case RIE_CMD_SWITCH_DESKTOP_PREV:
    case RIE_CMD_SWITCH_DESKTOP_NEXT:
        rie_switch_desktop(pager, cmd);
        break;
    case RIE_CMD_TILE_CURRENT_DESKTOP:
        pager->render = 1;
        (void) rie_windows_tile(pager, pager->current_desktop);
        break;
    }
}


void
rie_pager_delete(rie_t *pager, int final)
{
    rie_event_cleanup(pager);
    rie_skin_delete(pager->skin);
    rie_gfx_delete(pager->gfx);
    rie_control_delete(pager->ctl, final);
    rie_conf_cleanup(&pager->cfg->meta, pager->cfg);

    if (final) {
        rie_xcb_delete(pager->xcb);
        rie_log_delete(pager->log);
    }

    free(pager->cfg);
    free(pager);
}


void
rie_usage(char *p)
{
    fprintf(stdout, "Usage: %s [-h] [-v[v]] [-c <config>]"
                    " [-w] [-l] [-m <sockpath> <msg>]\n", p);
}


void
rie_version(int verbose)
{
    if (!verbose) {
        fprintf(stdout, "rieman "RIEMAN_VERSION"\n");
        return;
    }

    fprintf(stdout, "Rieman " RIEMAN_VERSION
" Copyright (c) 2017-2022  Vladimir Homutov\n"
"\n"
"This program comes with ABSOLUTELY NO WARRANTY.\n"
"This is free software, and you are welcome to redistribute it\n"
"under certain conditions; see the file COPYING for details.\n\n"

"Configured options:\n"
                    " · prefix:  \"" RIE_PREFIX "\"\n"
                    " · bindir:  \"" RIE_BINDIR  "\"\n"
                    " · mandir:  \"" RIE_MANDIR  "\"\n"
                    " · docdir:  \"" RIE_DOCDIR  "\"\n"
                    " · datadir: \"" RIE_DATADIR  "\"\n"
#if defined(RIE_DEBUG)
                    " · debug:   yes\n"
#else
                    " · debug:   no\n"
#endif

#if defined(RIE_TESTS)
                    " · tests:   yes\n"
#else
                    " · tests:   no\n"
#endif

                    "\nSources used:\n"
#if defined(RIE_REV)
                    " · git revision: \"" RIE_REV  "\"\n"
#else
                    " · tarball with version: \"" RIEMAN_VERSION  "\"\n"
#endif
                    "\nBuilt with:\n"
                    " · compiler: \"" RIE_CC "\"\n"
#if defined(__VERSION__)
                    " · compiler version: \"" __VERSION__ "\"\n"
#endif
                    " · build host: \"" RIE_BUILD_OS "\"\n"
                    "\nCompilation flags:\n"
                    " · CFLAGS: \"" RIE_CFLAGS "\"\n"
                    " · LDLAGS: \"" RIE_LDFLAGS "\"\n"
           "\n");
}


static void
rie_signal_handler(int sig_num)
{
    switch (sig_num) {
    case SIGTERM:
    case SIGINT:   /* CTRL-C */
        rie_quit = 1;
        break;
    case SIGUSR1:
        rie_reload = 1;
        break;
    default:
        /* should never happen */
        break;
    }
}


static int
rie_init_signals(sigset_t *ss)
{
    int  rc;

    struct sigaction  sa;

    sa.sa_flags = 0;
    sa.sa_handler = rie_signal_handler;

    rc = sigaction(SIGINT, &sa, NULL);
    if (-1 == rc) {
        rie_log_error0(errno, "sigaction()");
        return RIE_ERROR;
    }

    rc = sigaction(SIGTERM, &sa, NULL);
    if (-1 == rc) {
        rie_log_error0(errno, "sigaction()");
        return RIE_ERROR;
    }

    rc = sigaction(SIGUSR1, &sa, NULL);
    if (-1 == rc) {
        rie_log_error0(errno, "sigaction()");
        return RIE_ERROR;
    }

    /* block all signals */
    rc = sigfillset(ss);
    if (-1 == rc) {
        rie_log_error0(errno, "sigfillset()");
        return RIE_ERROR;
    }

    rc = sigprocmask(SIG_BLOCK, ss, NULL);
    if (-1 == rc) {
        rie_log_error0(errno, "sigprocmask()");
        return RIE_ERROR;
    }

    /* prepare signal mask for pselect() */

    rc = sigdelset(ss, SIGTERM);
    if (-1 == rc) {
        rie_log_error0(errno, "sigdelset()");
        return RIE_ERROR;
    }

    rc = sigdelset(ss, SIGINT);
    if (-1 == rc) {
        rie_log_error0(errno, "sigdelset()");
        return RIE_ERROR;
    }

    rc = sigdelset(ss, SIGUSR1);
    if (-1 == rc) {
        rie_log_error0(errno, "sigdelset()");
        return RIE_ERROR;
    }

    return RIE_OK;
}


rie_t *
rie_pager_new(char *cfile, rie_log_t *log)
{
    rie_t           *pager;
    rie_settings_t  *cfg;

    cfg = malloc(sizeof(rie_settings_t));
    if (cfg == NULL) {
        return NULL;
    }

    pager = malloc(sizeof(rie_t));
    if (pager == NULL) {
        free(cfg);
        return NULL;
    }

    rie_memzero(cfg, sizeof(rie_settings_t));
    rie_memzero(pager, sizeof(rie_t));

    cfg->meta.version_min = 10;
    cfg->meta.version_max = 12;
    cfg->meta.spec = rie_conf;

    rie_log("using configuration file '%s'", cfile);

    if (rie_conf_load(cfile, &cfg->meta, cfg) != RIE_OK) {
        rie_log_error(0, "configuration load failed");
        free(cfg);
        free(pager);
        return NULL;
    }

    cfg->meta.conf_file = cfile;

    pager->cfg = cfg;
    pager->log = log;

#if defined(RIE_TESTS)
    pagerp = pager;
#endif

    return pager;
}


int
main(int argc, char *argv[])
{
    int         i;
    char       *cfile, *logfile, *sockpath, *msg;
    rie_t      *pager;
    sigset_t    sigmask;
    rie_log_t  *log;

    char conf_file[FILENAME_MAX];

    cfile = NULL;
    logfile = NULL;
    msg = NULL;
    sockpath = NULL;

    if (argc > 1) {

        for (i = 1; i < argc; i++) {
            if (strcmp(argv[i], "-h") == 0) {
                rie_usage(argv[0]);
                exit(EXIT_SUCCESS);
            }

            if (strcmp(argv[i], "-v") == 0) {
                rie_version(0);
                exit(EXIT_SUCCESS);
            }

            if (strcmp(argv[i], "-vv") == 0) {
                rie_version(1);
                exit(EXIT_SUCCESS);
            }

            if (strcmp(argv[i], "-l") == 0) {
                if (argc < (i + 2)) {
                    fprintf(stderr, "option \"-l\" requires argument\n");
                    rie_usage(argv[0]);
                    exit(EXIT_FAILURE);
                }
                logfile = argv[i + 1];
                i++;
                continue;
            }

            if (strcmp(argv[i], "-c") == 0) {

                if (argc < (i + 2)) {
                    fprintf(stderr, "option \"-c\" requires argument\n");
                    rie_usage(argv[0]);
                    exit(EXIT_FAILURE);
                }

                cfile = argv[i + 1];
                i++;
                continue;
            }

            if (strcmp(argv[i], "-w") == 0) {
                rie_withdraw = 1;
                continue;
            }

            if (strcmp(argv[i], "-m") == 0) {

                if (argc < (i + 3)) {
                    fprintf(stderr, "option \"-m\" requires 2 arguments\n");
                    rie_usage(argv[0]);
                    exit(EXIT_FAILURE);
                }

                sockpath = argv[i + 1];
                msg = argv[i + 2];

                i += 2;

                continue;
            }

            fprintf(stderr, "Unknown argument \"%s\"\n", argv[i]);
            rie_usage(argv[0]);
            exit(EXIT_FAILURE);
        }
    }

    if (msg) {
        if (rie_control_send_message(sockpath, msg) != RIE_OK) {
            exit(EXIT_FAILURE);
        }

        exit(EXIT_SUCCESS);
    }

    log = rie_log_new(logfile);
    if (log == NULL) {
        fprintf(stderr, "failed to create log\n");
        exit(EXIT_FAILURE);
    }

    rie_log("rieman ver.%s (%s) started...", RIEMAN_VERSION, RIE_REV);

    if (cfile == NULL) {
        if (rie_locate_config(&conf_file, "rieman.conf") != RIE_OK) {
            fprintf(stderr, "No usable configuration file found, exiting\n");
            goto failed;
        }
        cfile = conf_file;
    }

    pager = rie_pager_new(cfile, log);
    if (pager == NULL) {
        goto failed;
    }

    if (rie_withdraw) {
        /* commandline overrides any default or config file */
        pager->cfg->withdrawn = 1;
    }

    if (rie_init_signals(&sigmask) != RIE_OK) {
        goto failed;
    }

    if (rie_pager_init(pager, NULL) != RIE_OK) {
        fprintf(stderr, "configuration failed to initialize, exiting\n");
        goto failed;
    }

    rie_log(" *** rieman is running with PID %d ***", (int) getpid());

    rie_log(" *** starting event loop ***");

    return rie_event_loop(pager, &sigmask);

failed:

    fprintf(stderr, ">> rieman failed to start due to fatal error\n");
    rie_log_dump();

    return EXIT_FAILURE;
}
